# Analysis

## **Finding range of hyperparameters**

Aim is to find the range of hyperparamentes for every model generated for each runtime.

Installing h20

In [4]:
!pip install h2o

    100% |████████████████████████████████| 121.0MB 200kB/s 
  Stored in directory: /root/.cache/pip/wheels/e9/92/d1/61cddd17defd8ba2a10400d2b3adf917377769610d904c279d
Successfully built h2o


Importing h2o and H2oAutoML libraries.

In [5]:
import h2o
from h2o.automl import H2OAutoML
import json
h2o.init()

Checking whether there is an H2O instance running at http://localhost:54321 ..... not found.
Attempting to start a local H2O server...
  Java Version: openjdk version "10.0.2" 2018-07-17; OpenJDK Runtime Environment (build 10.0.2+13-Ubuntu-1ubuntu0.18.04.4); OpenJDK 64-Bit Server VM (build 10.0.2+13-Ubuntu-1ubuntu0.18.04.4, mixed mode)
  Starting server from /usr/local/lib/python3.6/dist-packages/h2o/backend/bin/h2o.jar
  Ice root: /tmp/tmpr33jc8cm
  JVM stdout: /tmp/tmpr33jc8cm/h2o_unknownUser_started_from_python.out
  JVM stderr: /tmp/tmpr33jc8cm/h2o_unknownUser_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321 ... successful.


--------------------------  ---------------------------------------------------
H2O cluster uptime:         02 secs
H2O cluster timezone:       Etc/UTC
H2O data parsing timezone:  UTC
H2O cluster version:        3.24.0.2
H2O cluster version age:    9 days
H2O cluster name:           H2O_from_python_unknownUser_qdc269
H2O cluster total nodes:    1
H2O cluster free memory:    3 Gb
H2O cluster total cores:    2
H2O cluster allowed cores:  2
H2O cluster status:         accepting new members, healthy
H2O connection url:         http://127.0.0.1:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         Amazon S3, XGBoost, Algos, AutoML, Core V3, Core V4
Python version:             3.6.7 final
--------------------------  ---------------------------------------------------

Importing numpy and pandas libraries.

In [0]:
import numpy as np
import pandas as pd

**For XRT:**importing csv files into the dataframe.

In [0]:
df_xrt = h2o.import_file("XRT.csv - Sheet1.csv")

Parse progress: |█████████████████████████████████████████████████████████| 100%


Printing 5 rows of the dataframe.

In [0]:

df_xrt.head()

model_id stopping_rounds max_runtime_secs stopping_tolerance max_after_balance_size ntrees max_depth min_rows nbins nbins_top_level nbins_cats seed sample_rate col_sample_rate_per_tree col_sample_rate_change_per_level min_split_improvement mtries 
 
 
 XRT_1_AutoML_20190421_232600 0 0 0.00809803 5 50 20 1 20 1024 1024 -1.63149e+18 0.632 1 1 1e-05 -1 
 XRT_1_AutoML_20190421_232600 0 0 0.00809803 5 50 20 1 20 1024 1024 -1.63149e+18 0.632 1 1 1e-05 -1 
 XRT_1_AutoML_20190421_232600 0 0 0.00809803 5 50 20 1 20 1024 1024 -1.63149e+18 0.632 1 1 1e-05 -1 
 XRT_1_AutoML_20190421_232600 0 0 0.00809803 5 50 20 1 20 1024 1024 -1.63149e+18 0.632 1 1 1e-05 -1 
 XRT_1_AutoML_20190421_232600 0 0 0.00809803 5 50 20 1 20 1024 1024 -1.63149e+18 0.632 1 1 1e-05 -1 
 XRT_1_AutoML_20190421_233433 0 0 0.00809803 5 48 20 1 20 1024 1024 3.87543e+18 0.632 1 1 1e-05 -1 
 XRT_1_AutoML_20190421_233433 0 0 0.00809803 5 48 20 1 20 1024 1024 3.87543e+18 0.632 1 1 1e-05 -1 
 XRT_1_AutoML_20190421_233433 0 0 0.00809803 5 48 20 1 20 1024 1024 3.87543e+18 0.632 1 1 1e-05 -1 
 XRT_1_AutoML_20190421_233433 0 0 0.00809803 5 48 20 1 20 1024 1024 3.87543e+18 0.632 1 1 1e-05 -1 
 XRT_1_AutoML_20190421_234613 0 0 0.00809803 5 45 20 1 20 1024 1024 -6.72243e+18 0.632 1 1 1e-05 -1

Finding the min and max value for each column for integer and real values.

In [0]:
ranges = []
for key,val in df_xrt.types.items():
   print(key,val)
   if val == 'int':
        ranges.append([key,[df_xrt[key].min(),df_xrt[key].max()]])
   if val == 'real':
        ranges.append([key,[df_xrt[key].min(),df_xrt[key].max()]])

model_id enum
stopping_rounds int
max_runtime_secs int
stopping_tolerance real
max_after_balance_size int
ntrees int
max_depth int
min_rows int
nbins int
nbins_top_level int
nbins_cats int
seed int
sample_rate real
col_sample_rate_per_tree int
col_sample_rate_change_per_level int
min_split_improvement real
mtries int


Printing the datatype for each column.

In [0]:
df_gbm.dtypes

model_id                            object
stopping_rounds                      int64
max_runtime_secs                     int64
stopping_tolerance                  object
max_after_balance_size               int64
ntrees                               int64
max_depth                            int64
min_rows                             int64
nbins                                int64
nbins_top_level                      int64
nbins_cats                           int64
seed                                object
sample_rate                         object
col_sample_rate_per_tree            object
col_sample_rate_change_per_level    object
min_split_improvement               object
mtries                               int64
dtype: object

Range of each hyperparameter.

In [0]:
ranges

[['stopping_rounds', [0.0, 0.0]],
 ['max_runtime_secs', [0.0, 0.0]],
 ['stopping_tolerance', [0.008098029, 0.008098029]],
 ['max_after_balance_size', [5.0, 5.0]],
 ['ntrees', [43.0, 50.0]],
 ['max_depth', [20.0, 20.0]],
 ['min_rows', [1.0, 1.0]],
 ['nbins', [20.0, 20.0]],
 ['nbins_top_level', [1024.0, 1024.0]],
 ['nbins_cats', [1024.0, 1024.0]],
 ['seed', [-6.72243e+18, 6.359377122478104e+18]],
 ['sample_rate', [0.632, 0.632000029]],
 ['col_sample_rate_per_tree', [1.0, 1.0]],
 ['col_sample_rate_change_per_level', [1.0, 1.0]],
 ['min_split_improvement', [1e-05, 1e-05]],
 ['mtries', [-1.0, -1.0]]]

**For GBM:**importing csv files into the dataframe.

In [0]:
df_gbm = h2o.import_file("GBM.csv")

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [0]:
df_gbm.head()

model_id max_runtime_secs max_after_balance_size seed 
 
 
 GBM_4_AutoML_20190421_233433 5 0 -7.13e+18 
 GBM_2_AutoML_20190421_232600 5 0 -3.52063e+18 
 GBM_1_AutoML_20190421_233433 5 0 -7.12673e+18 
 GBM_2_AutoML_20190421_233433 5 0 -7.12673e+18 
 GBM_3_AutoML_20190421_234613 5 0 7.83787e+18 
 GBM_1_AutoML_20190421_234613 5 0 7.83787e+18 
 GBM_4_AutoML_20190421_234613 5 0 7.83787e+18 
 GBM_1_AutoML_20190421_232600 5 0 -3.52063e+18 
 GBM_2_AutoML_20190421_234613 5 0 7.83787e+18 
 GBM_3_AutoML_20190421_233433 5 0 -7.12673e+18

Finding the min and max value for each column for integer and real values.

In [0]:
ranges = []
for key,val in df_gbm.types.items():
   print(key,val)
   if val == 'int':
        ranges.append([key,[df_gbm[key].min(),df_gbm[key].max()]])
   if val == 'real':
        ranges.append([key,[df_gbm[key].min(),df_gbm[key].max()]])

model_id enum
max_runtime_secs int
max_after_balance_size int
seed int


Range of each hyperparameter. 

In [0]:
ranges

[['max_runtime_secs', [5.0, 5.0]],
 ['max_after_balance_size', [0.0, 0.0]],
 ['seed', [-8.27849e+18, 7.83787362635869e+18]]]

**For Deeplearning:**importing csv files into the dataframe.

In [0]:
df_dl = h2o.import_file("DL.csv")

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [0]:
df_dl.head()

model_id stopping_rounds max_runtime_secs stopping_tolerance max_after_balance_size seed tweedie_power quantile_alpha huber_alpha epochs train_samples_per_iteration target_ratio_comm_to_comp rho epsilon rate rate_decay momentum_start momentum_ramp momentum_stable input_dropout_ratio l1 l2 max_w2 initial_weight_scale score_interval score_training_samples score_validation_samples score_duty_cycle classification_stop regression_stop average_activation sparsity_beta max_categorical_features elastic_averaging_regularization 
 
 
 DeepLearning_1_AutoML_20190421_233433 0 0 0.00809803 5 -3.61794e+18 1.5 0.5 0.9 10.3488 -2 0.05 0.99 1e-06 0.005 1 0 1e+06 0 0 0 0 3.4e+38 1 5 10000 0 0.1 0 1e-06 0 0 2.14748e+09 0.001 
 DeepLearning_1_AutoML_20190421_232600 0 0 0.00809803 5 4.11e+18 1.5 0.5 0.9 10.3423 -2 0.05 0.99 1e-06 0.005 1 0 1e+06 0 0 0 0 3.4e+38 1 5 10000 0 0.1 0 1e-06 0 0 2.14748e+09 0.001 
 DeepLearning_1_AutoML_20190421_234613 0 0 0.00809803 5 1.95657e+18 1.5 0.5 0.9 10.3546 -2 0.05 0.99 1e-06 0.005 1 0 1e+06 0 0 0 0 3.4e+38 1 5 10000 0 0.1 0 1e-06 0 0 2.14748e+09 0.001 
 DeepLearning_grid_1_AutoML_20190421_233433_model_1 0 0 0.00809803 5 1.05538e+18 1.5 0.5 0.9 4.59518 -2 0.05 0.95 1e-06 0.005 1 0 1e+06 0 0.05 0 0 3.4e+38 1 5 10000 0 0.1 0 1e-06 0 0 2.14748e+09 0.001 
 DeepLearning_grid_1_AutoML_20190421_234613_model_2 0 0 0.00809803 5 8.80576e+18 1.5 0.5 0.9 0.47708 -2 0.05 0.99 1e-06 0.005 1 0 1e+06 0 0.15 0 0 3.4e+38 1 5 10000 0 0.1 0 1e-06 0 0 2.14748e+09 0.001 
 DeepLearning_grid_1_AutoML_20190421_232600_model_1 0 0 0.00809803 5 6.01537e+18 1.5 0.5 0.9 0.297593 -2 0.05 0.9 1e-06 0.005 1 0 1e+06 0 0.1 0 0 3.4e+38 1 5 10000 0 0.1 0 1e-06 0 0 2.14748e+09 0.001 
 DeepLearning_grid_1_AutoML_20190421_233433_model_4 0 0 0.00809803 5 -4.73713e+18 1.5 0.5 0.9 0.360679 -2 0.05 0.99 1e-06 0.005 1 0 1e+06 0 0.1 0 0 3.4e+38 1 5 10000 0 0.1 0 1e-06 0 0 2.14748e+09 0.001 
 DeepLearning_grid_1_AutoML_20190421_233433_model_2 0 0 0.00809803 5 1.59285e+18 1.5 0.5 0.9 0.332743 -2 0.05 0.95 1e-06 0.005 1 0 1e+06 0 0.15 0 0 3.4e+38 1 5 10000 0 0.1 0 1e-06 0 0 2.14748e+09 0.001 
 DeepLearning_grid_1_AutoML_20190421_234613_model_4 0 0 0.00809803 5 -8.67337e+18 1.5 0.5 0.9 0.210899 -2 0.05 0.9 1e-06 0.005 1 0 1e+06 0 0.2 0 0 3.4e+38 1 5 10000 0 0.1 0 1e-06 0 0 2.14748e+09 0.001 
 DeepLearning_grid_1_AutoML_20190421_233433_model_5 0 0 0.00809803 5 -3.99495e+18 1.5 0.5 0.9 0.694865 -2 0.05 0.9 1e-06 0.005 1 0 1e+06 0 0 0 0 3.4e+38 1 5 10000 0 0.1 0 1e-06 0 0 2.14748e+09 0.001

In [0]:
ranges = []
for key,val in df_dl.types.items():
   print(key,val)
   if val == 'int':
        ranges.append([key,[df_dl[key].min(),df_dl[key].max()]])
   if val == 'real':
        ranges.append([key,[df_dl[key].min(),df_dl[key].max()]])

model_id enum
stopping_rounds int
max_runtime_secs int
stopping_tolerance real
max_after_balance_size int
seed int
tweedie_power real
quantile_alpha real
huber_alpha real
epochs real
train_samples_per_iteration int
target_ratio_comm_to_comp real
rho real
epsilon real
rate real
rate_decay int
momentum_start int
momentum_ramp int
momentum_stable int
input_dropout_ratio real
l1 int
l2 int
max_w2 real
initial_weight_scale int
score_interval int
score_training_samples int
score_validation_samples int
score_duty_cycle real
classification_stop int
regression_stop real
average_activation int
sparsity_beta int
max_categorical_features int
elastic_averaging_regularization real


In [0]:
ranges

[['stopping_rounds', [0.0, 0.0]],
 ['max_runtime_secs', [0.0, 0.0]],
 ['stopping_tolerance', [0.008098029, 0.008098029]],
 ['max_after_balance_size', [5.0, 5.0]],
 ['seed', [-9.09605e+18, 8.805764935126581e+18]],
 ['tweedie_power', [1.5, 1.5]],
 ['quantile_alpha', [0.5, 0.5]],
 ['huber_alpha', [0.9, 0.9]],
 ['epochs', [0.084333401, 31.10764099]],
 ['train_samples_per_iteration', [-2.0, -2.0]],
 ['target_ratio_comm_to_comp', [0.05, 0.05]],
 ['rho', [0.9, 0.99]],
 ['epsilon', [1e-06, 1e-06]],
 ['rate', [0.005, 0.005]],
 ['rate_decay', [1.0, 1.0]],
 ['momentum_start', [0.0, 0.0]],
 ['momentum_ramp', [1000000.0, 1000000.0]],
 ['momentum_stable', [0.0, 0.0]],
 ['input_dropout_ratio', [0.0, 0.2]],
 ['l1', [0.0, 0.0]],
 ['l2', [0.0, 0.0]],
 ['max_w2', [3.4e+38, 3.4e+38]],
 ['initial_weight_scale', [1.0, 1.0]],
 ['score_interval', [5.0, 5.0]],
 ['score_training_samples', [10000.0, 10000.0]],
 ['score_validation_samples', [0.0, 0.0]],
 ['score_duty_cycle', [0.1, 0.1]],
 ['classification_stop', 

**For DRF:**importing csv files into the dataframe.

In [0]:
df_drf = h2o.import_file("DRF.csv")

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [0]:
df_drf.head()

model_id stopping_rounds max_runtime_secs stopping_tolerance max_after_balance_size ntrees max_depth min_rows nbins nbins_top_level nbins_cats seed sample_rate col_sample_rate_per_tree col_sample_rate_change_per_level min_split_improvement mtries 
 
 
 DRF_1_AutoML_20190421_232600 0 0 0.00809803 5 50 20 1 20 1024 1024 6.45958e+18 0.632 1 1 1e-05 -1 
 DRF_1_AutoML_20190421_233433 0 0 0.00809803 5 50 20 1 20 1024 1024 6.45958e+18 0.632 1 1 1e-05 -1 
 DRF_1_AutoML_20190421_234613 0 0 0.00809803 5 50 20 1 20 1024 1024 6.45958e+18 0.632 1 1 1e-05 -1 
 DRF_1_AutoML_20190421_232600 0 0 0.00809803 5 50 20 1 20 1024 1024 6.45958e+18 0.632 1 1 1e-05 -1 
 DRF_1_AutoML_20190421_233433 0 0 0.00809803 5 50 20 1 20 1024 1024 6.45958e+18 0.632 1 1 1e-05 -1 
 DRF_1_AutoML_20190421_234613 0 0 0.00809803 5 50 20 1 20 1024 1024 7.5491e+18 0.632 1 1 1e-05 -1 
 DRF_1_AutoML_20190422_000108 0 0 0.00809803 5 50 20 1 20 1024 1024 7.5491e+18 0.632 1 1 1e-05 -1 
 DRF_1_AutoML_20190421_232600 0 0 0.00809803 5 50 20 1 20 1024 1024 7.5491e+18 0.632 1 1 1e-05 -1 
 DRF_1_AutoML_20190421_232600 0 0 0.00809803 5 50 20 1 20 1024 1024 7.5491e+18 0.632 1 1 1e-05 -1 
 DRF_1_AutoML_20190421_233433 0 0 0.00809803 5 41 20 1 20 1024 1024 -9.12e+18 0.632 1 1 1e-05 -1

In [0]:
ranges = []
for key,val in df_drf.types.items():
   print(key,val)
   if val == 'int':
        ranges.append([key,[df_drf[key].min(),df_drf[key].max()]])
   if val == 'real':
        ranges.append([key,[df_drf[key].min(),df_drf[key].max()]])

model_id enum
stopping_rounds int
max_runtime_secs int
stopping_tolerance real
max_after_balance_size int
ntrees int
max_depth int
min_rows int
nbins int
nbins_top_level int
nbins_cats int
seed int
sample_rate real
col_sample_rate_per_tree int
col_sample_rate_change_per_level int
min_split_improvement real
mtries int


In [0]:
ranges

[['stopping_rounds', [0.0, 0.0]],
 ['max_runtime_secs', [0.0, 0.0]],
 ['stopping_tolerance', [0.008098029, 0.008098029]],
 ['max_after_balance_size', [5.0, 5.0]],
 ['ntrees', [41.0, 50.0]],
 ['max_depth', [20.0, 20.0]],
 ['min_rows', [1.0, 1.0]],
 ['nbins', [20.0, 20.0]],
 ['nbins_top_level', [1024.0, 1024.0]],
 ['nbins_cats', [1024.0, 1024.0]],
 ['seed', [-9.12e+18, 7.549101339683506e+18]],
 ['sample_rate', [0.632000029, 0.632000029]],
 ['col_sample_rate_per_tree', [1.0, 1.0]],
 ['col_sample_rate_change_per_level', [1.0, 1.0]],
 ['min_split_improvement', [1e-05, 1e-05]],
 ['mtries', [-1.0, -1.0]]]

**For GLM:**importing csv files into the dataframe.

In [0]:
df_glm = h2o.import_file("GLM.csv")

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [0]:
df_glm.head()

model_id max_runtime_secs max_after_balance_size seed max_iterations tweedie_variance_power tweedie_link_power 
 
 
 GLM_grid_1_AutoML_20190421_232600_model_1 0 5 6.56e+17 300 0 1 
 GLM_grid_1_AutoML_20190421_232600_model_1 0 5 6.56e+17 300 0 1 
 GLM_grid_1_AutoML_20190421_232600_model_1 0 5 6.56e+17 300 0 1 
 GLM_grid_1_AutoML_20190421_232600_model_1 0 5 6.56e+17 300 0 1 
 GLM_grid_1_AutoML_20190421_232600_model_1 0 5 6.56e+17 300 0 1 
 GLM_grid_1_AutoML_20190421_233433_model_1 0 5 6.23814e+17 300 0 1 
 GLM_grid_1_AutoML_20190421_233433_model_1 0 5 6.23814e+17 300 0 1 
 GLM_grid_1_AutoML_20190421_233433_model_1 0 5 6.23814e+17 300 0 1 
 GLM_grid_1_AutoML_20190421_233433_model_1 0 5 6.23814e+17 300 0 1 
 GLM_grid_1_AutoML_20190421_233433_model_1 0 5 1.92983e+18 300 0 1

In [0]:
ranges = []
for key,val in df_glm.types.items():
   print(key,val)
   if val == 'int':
        ranges.append([key,[df_glm[key].min(),df_glm[key].max()]])
   if val == 'real':
        ranges.append([key,[df_glm[key].min(),df_glm[key].max()]])

model_id enum
max_runtime_secs int
max_after_balance_size int
seed int
max_iterations int
tweedie_variance_power int
tweedie_link_power int


In [0]:
ranges

[['max_runtime_secs', [0.0, 0.0]],
 ['max_after_balance_size', [5.0, 5.0]],
 ['seed', [-8.93342e+18, 1.929825150468839e+18]],
 ['max_iterations', [300.0, 300.0]],
 ['tweedie_variance_power', [0.0, 0.0]],
 ['tweedie_link_power', [1.0, 1.0]]]

# Compare the range of values across the models for different hyperparameters :


> XRT, GLM and GBM have common hyperparameters i.e. max_runtime_secs, max_after_balance_size, seed.

> Value of hyperparameter Max_depth is same for GBM and DRF.

> DRF, XRT and Deep Learning have common hyperparameters i.e. stopping_rounds, max_runtime_secs, stopping_tolerance, max_after_balance_size, Seed.

> Stopping_tolerance is present is all the algorithms except GLM and GBM and its value is 0.008098029 for all models.

> Seed is common in all algorithms and its range varies in all algorithms. Value of seed varies as follows:
  



>   For GBM: [-6.72243e+18, 6.359377122478104e+18]

 > For DeepLearning: [-8.27849e+18, 7.83787362635869e+18]

  > For DRF: [-9.09605e+18, 8.805764935126581e+18]

 >  For GLM: [-9.12e+18, 7.549101339683506e+18]]

  > For GBM: [-8.93342e+18, 1.929825150468839e+18]


> Ntrees are for XRT and DRF only and the valuea varies from 41.0 to 50.0

>  For DeepLearning input_dropout_ratio varies from [0.0, 0.2]

> For DRF sample_rate varies from [0.632, 0.632000029]



# Grid Search for finding Important hyperparameters

**For GBM:**importing csv files into the dataframe.

In [0]:
from h2o.estimators.gbm import H2OGradientBoostingEstimator
from h2o.grid.grid_search import H2OGridSearch

In [0]:
# Load data into H2O
df2 = h2o.import_file("bottle.csv")
x = df2.columns
y = "T_degC"
x.remove(y)

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [0]:
split = df2.split_frame(seed = 1)
train = split[0]
valid = split[1]

In [0]:
gbm_params = {'max_runtime_secs', [5.0, 5.0],
 'max_after_balance_size', [0.0, 0.0],
 'seed', [-8.27849e+18, 7.83787362635869e+18]}

In [0]:
gbm_search = H2OGridSearch(model=H2OGradientBoostingEstimator,
                          grid_id='gbm_search',
                          hyper_params=gbm_params)

In [0]:
gbm_search.train(x=x, y=y,
                training_frame=train,
                validation_frame=valid,
                ntrees=50,
                seed=1)

gbm Grid Build progress: |████████████████████████████████████████████████| 100%


In [0]:
gbm_best = gbm_search.get_grid(sort_by='mse', decreasing=True)
gbm_best

    huber_alpha learn_rate max_after_balance_size max_depth min_rows nbins  \
0           0.9      0.001                    5.0         3    100.0    20   
1           0.9      0.001                    5.0         3      5.0    20   
2           0.9      0.001                    5.0        17    100.0    20   
3           0.9      0.001                    5.0        17      5.0    20   
4           0.9        0.5                    5.0         3    100.0    20   
5           0.9        0.5                    5.0         3      5.0    20   
6           0.9        0.5                    5.0        17    100.0    20   
7           0.9        0.5                    5.0        17      5.0    20   

  nbins_top_level quantile_alpha    stopping_tolerance tweedie_power  \
0            1024            0.5  0.006802091999999999           1.5   
1            1024            0.5  0.006802091999999999           1.5   
2            1024            0.5  0.006802091999999999           1.5   
3        

In [0]:
best_gbm = gbm_best.models[0]

In [0]:
best_gbm

Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Machine
Model Key:  gbm_search_model_17


ModelMetricsRegression: gbm
** Reported on train data. **

MSE: 16.34846021587728
RMSE: 4.0433229175861385
MAE: 3.3150862641911285
RMSLE: 0.3713922305183845
Mean Residual Deviance: 16.34846021587728

ModelMetricsRegression: gbm
** Reported on validation data. **

MSE: 16.34957892052357
RMSE: 4.043461254979893
MAE: 3.317418108125699
RMSLE: 0.3705922642581966
Mean Residual Deviance: 16.34957892052357
Scoring History: 


timestamp            duration           number_of_trees    training_rmse    training_mae    training_deviance    validation_rmse    validation_mae    validation_deviance
--  -------------------  -----------------  -----------------  ---------------  --------------  -------------------  -----------------  ----------------  ---------------------
    2019-04-26 15:27:35  36 min 41.029 sec  0                  4.24381          3.48113         18.0099              4.24387            3.4835            18.0104
    2019-04-26 15:27:52  36 min 58.483 sec  5                  4.2233           3.46415         17.8362              4.22337            3.46652           17.8369
    2019-04-26 15:28:01  37 min  7.853 sec  8                  4.21104          3.45401         17.7329              4.21112            3.45637           17.7336
    2019-04-26 15:28:10  37 min 16.624 sec  11                 4.19882          3.44389         17.6301              4.19891            3.44626           17.6309
    2019-04-26 15:28:19  37 min 25.830 sec  14                 4.18664          3.43381         17.528               4.18674            3.43617           17.5288
    2019-04-26 15:28:28  37 min 34.443 sec  17                 4.1745           3.42375         17.4264              4.1746             3.42612           17.4273
    2019-04-26 15:28:36  37 min 42.752 sec  20                 4.16239          3.41372         17.3255              4.1625             3.41609           17.3264
    2019-04-26 15:28:45  37 min 51.319 sec  23                 4.15032          3.40373         17.2252              4.15044            3.40609           17.2261
    2019-04-26 15:28:54  38 min  0.263 sec  26                 4.13829          3.39376         17.1254              4.13841            3.39612           17.1264
    2019-04-26 15:29:03  38 min  9.594 sec  29                 4.12629          3.38382         17.0263              4.12642            3.38618           17.0273
    2019-04-26 15:29:12  38 min 18.856 sec  32                 4.11433          3.37391         16.9277              4.11446            3.37627           16.9288
    2019-04-26 15:29:21  38 min 27.544 sec  35                 4.10241          3.36404         16.8297              4.10253            3.36639           16.8308
    2019-04-26 15:29:30  38 min 36.363 sec  38                 4.09052          3.35419         16.7323              4.09065            3.35654           16.7334
    2019-04-26 15:29:39  38 min 45.557 sec  41                 4.07867          3.34437         16.6355              4.0788             3.34672           16.6366
    2019-04-26 15:29:47  38 min 53.995 sec  44                 4.06685          3.33458         16.5393              4.06699            3.33692           16.5404
    2019-04-26 15:29:56  39 min  2.617 sec  47                 4.05507          3.32482         16.4436              4.05521            3.32715           16.4447
    2019-04-26 15:30:05  39 min 11.411 sec  50                 4.04332          3.31509         16.3485              4.04346            3.31742           16.3496

Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
R_TEMP      532408896.0            1.0                  1.0
Cst_Cnt     0.0                    0.0                  0.0
Btl_Cnt     0.0                    0.0                  0.0
Sta_ID      0.0                    0.0                  0.0
Depthm      0.0                    0.0                  0.0
---         ---                    ---                  ---
DIC2        0.0                    0.0                  0.0
TA1         0.0                    0.0                  0.0
TA2         0.0                    0.0                  0.0
pH2         0.0                    0.0                  0.0
pH1         0.0                    0.0                  0.0


See the whole table with table.as_data_frame()


**For GLM**:importing csv files into the dataframe.

In [0]:
glm_params1 = { 'max_runtime_secs': [0.0, 0.0],
 #'max_after_balance_size': [5.0, 5.0],
 'seed': [-8.93342e+18, 1.929825150468839e+18],
 #'max_iterations': [300.0, 300.0],
 'tweedie_variance_power': [0.0, 0.0],
 'tweedie_link_power': [1.0, 1.0]}

In [0]:
from h2o.estimators import H2ORandomForestEstimator
from h2o.estimators import H2OGeneralizedLinearEstimator
from h2o.grid.grid_search import H2OGridSearch
# Train and validate a cartesian grid of GBMs
glm_grid1 = H2OGridSearch(model=H2OGeneralizedLinearEstimator,
                         grid_id='glm',
                         hyper_params=glm_params1)

In [22]:
# Load data into H2O
df2 = h2o.import_file("bottle.csv")
X = df2.columns
y = "T_degC"
X.remove(y)

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [23]:
glm_grid1.train(X,y, training_frame=df2)

glm Grid Build progress: |████████████████████████████████████████████████| 100%


In [25]:
glm_best = glm_grid1.get_grid(sort_by='mse', decreasing=True)
glm_best

           max_runtime_secs                  seed tweedie_link_power  \
0    1.7976931348623157E308   1929825150468838910                1.0   
1    1.7976931348623157E308  -8933420000000000000                1.0   

  tweedie_variance_power    model_ids                   mse  
0                    0.0  glm_model_3  0.005134416198101295  
1                    0.0  glm_model_1  0.005134416198101295  


In [26]:
best_glm = glm_best.models[0]
best_glm

Model Details
H2OGeneralizedLinearEstimator :  Generalized Linear Modeling
Model Key:  glm_model_3


ModelMetricsRegressionGLM: glm
** Reported on train data. **

MSE: 0.005134416198101295
RMSE: 0.07165484071646029
MAE: 0.02828093812584049
RMSLE: 0.011520900428912201
R^2: 0.9997149134585768
Mean Residual Deviance: 0.005134416198101295
Null degrees of freedom: 853899
Residual degrees of freedom: 853886
Null deviance: 15378761.7250122
Residual deviance: 4384.277991558696
AIC: -2078287.513486648
Scoring History: 


timestamp            duration    iterations    negative_log_likelihood    objective
--  -------------------  ----------  ------------  -------------------------  -----------
    2019-04-26 21:37:40  0.000 sec   0             1.53788e+07                17.7817

**For DRF:**importing csv files into the dataframe.

In [0]:
drf_params1 = { 'mtries': [-1.0, -1.0]}

In [0]:
from h2o.estimators import H2ORandomForestEstimator
from h2o.estimators import H2OGeneralizedLinearEstimator
from h2o.grid.grid_search import H2OGridSearch
# Train and validate a cartesian grid of GBMs
drf_grid1 = H2OGridSearch(model=H2ORandomForestEstimator,
                         grid_id='drf',
                         hyper_params=drf_params1)

In [29]:
# Load data into H2O
df2 = h2o.import_file("bottle.csv")
X = df2.columns
y = "T_degC"
X.remove(y)

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [30]:
drf_grid1.train(X,y, training_frame=df2)

drf Grid Build progress: |████████████████████████████████████████████████| 100%


In [31]:
drf_best = drf_grid1.get_grid(sort_by='mse', decreasing=True)
drf_best

    mtries    model_ids                   mse
0       -1  drf_model_1  6.898143442307281E-4


In [32]:
best_drf = drf_best.models[0]
best_drf

Model Details
H2ORandomForestEstimator :  Distributed Random Forest
Model Key:  drf_model_1


ModelMetricsRegression: drf
** Reported on train data. **

MSE: 0.0006898143442307281
RMSE: 0.026264316938209683
MAE: 0.004679474804337884
RMSLE: 0.0020165709021871528
Mean Residual Deviance: 0.0006898143442307281
Scoring History: 


timestamp            duration           number_of_trees    training_rmse         training_mae           training_deviance
---  -------------------  -----------------  -----------------  --------------------  ---------------------  ---------------------
     2019-04-26 21:46:41  0.032 sec          0.0                nan                   nan                    nan
     2019-04-26 21:47:17  36.321 sec         1.0                0.04847395564593441   0.00547135685839257    0.0023497243759640165
     2019-04-26 21:47:40  59.191 sec         2.0                0.040801379848574286  0.004297173906708278   0.0016647525975476438
     2019-04-26 21:48:03  1 min 22.709 sec   3.0                0.038811960292638045  0.0041514901425557145  0.0015063682617573126
     2019-04-26 21:48:26  1 min 45.896 sec   4.0                0.03350684038637928   0.003911317624393318   0.0011227083526782974
---  ---                  ---                ---                ---                   ---                    ---
     2019-04-26 22:04:22  17 min 41.204 sec  46.0               0.027229651013508417  0.004808662997010338   0.00074145389431746
     2019-04-26 22:04:46  18 min  5.017 sec  47.0               0.02710265769762446   0.004740071056247393   0.0007345540542746024
     2019-04-26 22:05:06  18 min 25.007 sec  48.0               0.026706147374485775  0.00466912234045591    0.0007132183075877535
     2019-04-26 22:05:30  18 min 49.460 sec  49.0               0.02623281552799333   0.0046296343430460105  0.000688160610525728
     2019-04-26 22:05:49  19 min  8.048 sec  50.0               0.026264316938209683  0.004679474804337884   0.0006898143442307281


See the whole table with table.as_data_frame()
Variable Importances: 


variable    relative_importance     scaled_importance       percentage
----------  ----------------------  ----------------------  ----------------------
R_TEMP      250028336.0             1.0                     0.36289086438817764
R_SVA       146138896.0             0.5844893356407411      0.21210584023634024
R_POTEMP    114683520.0             0.45868209113706215     0.16645154053210537
R_SIGMA     57247208.0              0.2289628804312804      0.0830885375925143
STheta      44639352.0              0.1785371718827901      0.06478950862996635
---         ---                     ---                     ---
P_qual      0.017908815294504166    7.162714267115775e-11   2.5992835717591758e-11
TA2         0.0005668838275596499   2.267278327843809e-12   8.227745921998218e-13
DIC2        0.00030103951576165855  1.2040215944230359e-12  4.3692843714220737e-13
pH1         0.0001449681876692921   5.798070330288167e-13   2.10406675394172e-13
pH2         0.0                     0.0                     0.0


See the whole table with table.as_data_frame()


#Conclusion:
Found the range of hyperparamentes for every model generated for each runtime.

The H2O was run on the data set to obtain the parameters for the various algorithms for different run times.

Compare the range of values across the models for different hyperparameters.

The Grid Search technique was then performed on the GBM to find out the best hyperparameter for the best performing model.

# Contributions:
Percentage of code contributed individually for this project: 60%
Percentage of code borrowed from other sources: 40%

# Citations:
http://docs.h2o.ai/h2o/latest-stable/h2o-docs/grid-search.html

http://docs.h2o.ai/h2o/latest-stable/h2o-docs/faq/python.html

#License:
Copyright <2019> FNU Karan

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.